### PR TITLE
Feature/train

### PR DESCRIPTION
**Pull request by HLasse**

### CODE REVIEW

File: src/dfm/data/load.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
+"""Dataset loading script for Danish Foundation Models
+"""
```
Comment (saattrupdan) [inline src/dfm/data/load.py:2]: Remove newline

File: src/dfm/data/load.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+        dataset (str): string for Hugging Face dataset.
+
+    Returns:
+        DatasetDict: Dataset with train, test and validation split.
+    """
+
+    ds = load_dataset(dataset, streaming=True)
+
+    if "validation" and "test" not in ds:
+        train_test = ds["train"].train_test_split(test_size=0.1)
```
Comment (saattrupdan) [inline src/dfm/data/load.py:26]: Use random seed here, for reproducibility?

File: src/dfm/data/load.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        # Split the 10% test + valid in half test, half valid
+        test_valid = train_test["test"].train_test_split(test_size=0.5)
```
Comment (saattrupdan) [inline src/dfm/data/load.py:28]: Use random seed here, for reproducibility?

File: src/dfm/data/preprocess.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
+"""Data preprocessing script for Danish Foundation Models """
+from typing import List, Union
```
Comment (saattrupdan) [inline src/dfm/data/preprocess.py:2]: Missing newline after module docstring

File: src/dfm/data/preprocess.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+    ds = preprocess_dataset(ds, tokenizer)
+
+
+def preprocess_dataset(
+    dataset: DatasetDict,
+    tokenizer: Union[PreTrainedTokenizerFast, PreTrainedTokenizerBase],
+    num_proc: int = 4,
+    block_size: int = 512,
+):
```
Comment (saattrupdan) [inline src/dfm/data/preprocess.py:28]: Missing docstring.

File: src/dfm/data/preprocess.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+        batched=True,
+        batch_size=1000,
+        num_proc=num_proc,
+    )
+    dataset.shuffle()
+
+    return dataset
+
+
+def select_column_(dataset: Dataset, column: str):
```
Comment (saattrupdan) [inline src/dfm/data/preprocess.py:56]: Missing docstring.

File: src/dfm/data/preprocess.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+    return dataset[column]
+
+
+def tokenize_func(examples, tokenizer):
```
Comment (saattrupdan) [inline src/dfm/data/preprocess.py:60]: Missing docstring.

File: src/dfm/data/preprocess.py [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+    return tokenizer(examples["text"])
+
+
+def group_texts(examples, block_size: int):
```
Comment (saattrupdan) [inline src/dfm/data/preprocess.py:64]: Missing docstring.

File: src/dfm/data/preprocess.py [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+    # Concatenate all texts.
+    concatenated_examples = {k: sum(examples[k], []) for k in examples.keys()}
+    total_length = len(concatenated_examples[list(examples.keys())[0]])
+    # We drop the small remainder, we could add padding if the model supported it instead of this drop, you can
```
Comment (saattrupdan) [inline src/dfm/data/preprocess.py:68]: Add newline before comment.

File: src/dfm/data/preprocess.py [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+    # customize this part to your needs.
+    total_length = (total_length // block_size) * block_size
+    # Split by chunks of max_len.
```
Comment (saattrupdan) [inline src/dfm/data/preprocess.py:71]: Add newline before comment.

File: src/dfm/modelling/train.py [hunk 11]
```diff
@@ -0,0 +1,104 @@
+"""Training script for training masked language models with sweeps
+Inspiration: https://colab.research.google.com/github/wandb/examples/blob/master/colabs/pytorch/Organizing_Hyperparameter_Sweeps_in_PyTorch_with_W%26B.ipynb#scrollTo=r4VjKui20N3j
+
+https://colab.research.google.com/github/huggingface/notebooks/blob/master/examples/language_modeling_from_scratch.ipynb
+
+
```
Comment (saattrupdan) [inline src/dfm/modelling/train.py:6]: Remove double newline

File: src/dfm/modelling/train.py [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+    Trainer,
+    TrainingArguments,
+    AutoTokenizer,
+    DataCollatorForLanguageModeling,
+)
+from pathlib import Path
+from typing import List, Optional
+from datasets import interleave_datasets
+
+
```
Comment (saattrupdan) [inline src/dfm/modelling/train.py:24]: Remove double newline

File: src/dfm/modelling/train.py [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+    pretraining_config_path: str,
+    dataset_names: List[str],
+    interleave_probabilities: Optional[List[float]],
+):
+    """
+    # TODO:
+        # testes med reddit
+        # create configs for models (training + model)
+        # test 3 modeller t5, deberta v2, gpt
+
```
Comment (saattrupdan) [inline src/dfm/modelling/train.py:49]: Remove these TODO lines?

File: src/dfm/modelling/train.py [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+    # Configuration
+    pretrain_config = read_json(pretraining_config_path)
+
+    HF_config = AutoConfig.from_pretrained(model_name)
+    tokenizer = AutoTokenizer.from_pretrained(model_name)
+
+    # Load and preprocess datasets
+    datasets = [dfm_load_dataset(d) for d in dataset_names]
+    # might want to set num_proc to more than 4 in `preprocess_dataset``
```
Comment (saattrupdan) [inline src/dfm/modelling/train.py:69]: Add newline before comment.

File: src/dfm/modelling/train.py [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+    datasets = [preprocess_dataset(d, tokenizer=tokenizer) for d in datasets]
+    train_datasets = [d["train"] for d in datasets]
+    eval_datasets = [d["validation"] for d in datasets]
+    # Interleave
```
Comment (saattrupdan) [inline src/dfm/modelling/train.py:73]: Add newline before comment.

File: src/dfm/modelling/train.py [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+    train_datasets = interleave_datasets(
+        train_datasets, probabilities=interleave_probabilities
+    )
+    eval_datasets = interleave_datasets(
+        eval_datasets, probabilities=interleave_probabilities
+    )
+    # Data Collator
```
Comment (saattrupdan) [inline src/dfm/modelling/train.py:80]: Add newline before comment.

File: src/dfm/modelling/train.py [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+        # testes med reddit
+        # create configs for models (training + model)
+        # test 3 modeller t5, deberta v2, gpt
+
+    Training function with Weights & Biases logging.
+
+    Args:
+        model_name (str): Path to a model.
+        pretraining_config_path (str): Path to a pretraining config.
+        datasets (list(str)): List of dataset names from the Hugging Face hub.
```
Comment (saattrupdan) [inline src/dfm/modelling/train.py:55]: Wrong argument name.

File: src/dfm/data/preprocess.py [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+):
+
+    for key in dataset.keys():
+        cols = dataset[key].column_names
+        cols.remove("text")
+        dataset[key] = dataset[key].remove_columns(cols)
+        # select the text column
+        ##dataset[key] = dataset[key].map(
+        #    select_column_, input_columns="text", remove_columns=dataset[key].column_names
+        # )
```
Comment (saattrupdan) [inline src/dfm/data/preprocess.py:37]: Remove these lines?

File: src/dfm/data/preprocess.py [hunk 19]
```diff
@@ -0,0 +1,81 @@
...
+    )
+    group_texts_ = partial(group_texts, block_size=block_size)
+
+    dataset = dataset.map(
+        group_texts_,
+        batched=True,
+        batch_size=1000,
+        num_proc=num_proc,
+    )
+    dataset.shuffle()
```
Comment (saattrupdan) [inline src/dfm/data/preprocess.py:51]: This doesn't perform a proper shuffle as `dataset` is an iterable dataset, right? Maybe make that clear if it's the case.

File: src/dfm/data/preprocess.py [hunk 20]
```diff
@@ -0,0 +1,81 @@
...
+
+
+def tokenize_func(examples, tokenizer):
+    return tokenizer(examples["text"])
+
+
+def group_texts(examples, block_size: int):
+    # Concatenate all texts.
+    concatenated_examples = {k: sum(examples[k], []) for k in examples.keys()}
+    total_length = len(concatenated_examples[list(examples.keys())[0]])
```
Comment (saattrupdan) [inline src/dfm/data/preprocess.py:67]: Doesn't this load the dataset into memory, which we can't do?

**Review by saattrupdan — CHANGES_REQUESTED** [review]
Overall looks good! Consider collecting everything in a class, as several of the functions are just helper functions to other functions, and a class would just make it easier to organise when we're creating the overall pipeline.
